test/test_prototype_builtin_datasets.py

```diff
@@ -131,8 +130,31 @@ def scan(graph):
 
         dataset = datasets.load(dataset_mock.name, **config)
 
-        if not any(type(dp) is annotation_dp_type for dp in scan(traverse(dataset))):
-            raise AssertionError(f"The dataset doesn't contain a {annotation_dp_type.__name__}() datapipe.")
+        # The nodes are in reversed order, i.e the last datapipe in the dataset comes first
+        nodes = iter(scan(traverse(dataset)))
+
+        for node in nodes:
+            if isinstance(node, ShardingFilter):
+                break
+            elif isinstance(node, Shuffler):
+                raise AssertionError(
+                    "Found a `Shuffler` after a `ShardingFilter` in the graph. "
+                    "Please place the `dp = hint_shuffling(dp)` call before the `dp = hint_sharding(dp)` call."
+                )
+        else:
+            raise AssertionError(
+                "Found no `ShardingFilter` in the graph. "
+                "Please place a `dp = hint_sharding(dp)` call in the dataset implementation. "
+            )
+
+        for node in nodes:
+            if isinstance(node, Shuffler):
+                break
+        else:
+            raise AssertionError(
+                "Found no `Shuffler` in the graph. "
+                "Please place a `dp = hint_shuffling(dp)` call in the dataset implementation. "
+            )
```

Comment:
@NivekT @ejguan @VitalyFedyunin This or similar logic can probably reused for the warning we potentially wanted to emit in the data loader.